README.md
# merge-ics
Forked from https://github.com/tuxpeople/merge-ics

Added php wrapper to fetch various calendars and merge them together on the fly.


Please note, that this script currently only supports the VEVENT-tag. Nothing else (this means: no todo's etc...).


CHANGELOG
merge_ics 2.0 (2015-04-17)
==========================

* Forked from https://code.google.com/p/merge-ics/
* Added config file
* Hacked together a php wrapper to perform merging on the fly to server calendar clients

merge_ics 1.6 (2007-07-09)
==========================

* Wrote warning into README regarding the VEVENT 'thing'
* split up header into different fiels (dependencies, credits etc...)
* fixed bug with timezone (sunbird wants a VTIMEZONE section, now we have one)
* Added Time/Date to debug output


config.py
#############################
# Variables:                #
#############################

# Folder, where the existing *.ics files are. The trailing slash is neccessary. This directory (and its files) must be readable for the user which runs this script
CALDIR = '/var/www/izzl.org/caltest/work/cals/'

# Name (and path) where the merged ics-file should be written. This directory and file must be writeable for the user which runs this script
ICS_OUT = '/var/www/izzl.org/caltest/work/cal.ics'

# Name of this tool
MY_NAME = 'Merged Calendar for Jeff Vogt'

# Our domain (This and MY_NAME will be printed into the new ics, as 'generator' ;) )
MY_DOMAIN = 'izzl.org'

# a short, one word name for this tool, used in errormessages as identifier
MY_SHORTNAME = 'merge_ics.py'

# The Timezone for the new file. This is here, becaus Mozilla Sunbird 0.5 want it ;). It's the general Timezone of the file. Normaly, an entry has it one timezone.
OUR_TIMEZONE = 'US/Pacific'

# Name of the new calendar
CALENDARNAME = 'Merged Calendar for Jeff Vogt'

# do we need logging?
DEBUG = True

# if DEBUG=True, where to log the data? (note: logfile (if it already exists) or directory must be writable by the user which runs this script)
DEBUGFILE = '/var/www/izzl.org/caltest/work/merge_ics.log'


config.php
<?php
$install_dir = '/var/www/izzl.org/caltest';
$files = array(
	array('personal.ics','http://outlook.office365.com/owa/calendar/bc7ae11819b24abc8c3200bbe09a67a0@mheducation.com/7c719939459c402bb74e6ab84e2d283517443460057374595992/calendar.ics'),
	array('personal-limited.ics','http://outlook.office365.com/owa/calendar/bc7ae11819b24abc8c3200bbe09a67a0@mheducation.com/aee5d081a0654d7997e810184d887f341029223846656721814/calendar.ics -O /var/www/izzl.org/caltest/work/cals/personal-limited.ics'),
	array('work.ics','http://outlook.office365.com/owa/calendar/bc7ae11819b24abc8c3200bbe09a67a0@mheducation.com/da76ba5ea6ab409988fc4ace92c641ee11786315112047513355/calendar.ics'),
	array('google.ics','https://www.google.com/calendar/ical/jdvogt%40gmail.com/private-79c8339b9604831b19d9b3ea420d685b/basic.ics')
);
?>


merge_ics.py
#!/usr/bin/python
#
#      merge_ics.py:        This script will get all .ics files (iCalendar files, as
#                           specified in the RFC 2445 specification), read it and
#                           aggregate all events to a new .ics file. If one of the
#                           sourcefiles is not readable (or is not RFC 2445 compatible),
#                           it will be ignored.
# 
#      Original Author:     Thomas Deutsch <thomas@tuxpeople.org>
#      Current Maintainer:  Jeff Vogt (http://github.com/jvogt)
#
#      Version:             2.0(2015-04-17)
#       
#      License:             GPL v2
#       
#                           This program is free software; you can redistribute it and/or modify
#                           it under the terms of the GNU General Public License as published by
#                           the Free Software Foundation; either version 2 of the License, or
#                           (at your option) any later version.
#       
#                           This program is distributed in the hope that it will be useful, but
#                           WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
#                           or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
#                           for more details.
#       
#                           You should have received a copy of the GNU General Public License along
#                           with this program; if not, write to the Free Software Foundation, Inc.,
#                           51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA 
#       
#      Usage:               This script is made to be started by hand or via the php wrapper.
#                           It needs no parameters or options, but you should configure
#                           the variables in the config file before you run it ;)

from config import *
#############################
# Script:                   #
#############################

# We need some stuff
import glob, sys, datetime
from time import *

DEBUGMSG = 'Log started\n'

lt = localtime()
DEBUGMSG = DEBUGMSG + strftime("Tag/Monat/Jahr: %d/%m/%Y", lt) + '\n'
DEBUGMSG = DEBUGMSG + strftime("Stunde:Minute:Sekunde: %H:%M:%S", lt) + '\n'

DEBUGMSG = DEBUGMSG + 'import of glob and sys done\n'

# We need the iCalendar package from http://codespeak.net/icalendar/
import icalendar
from icalendar import Calendar, Event, Timezone
DEBUGMSG = DEBUGMSG + 'inport of icalendar done\n'

# Open the new calendarfile and adding the information about this script to it
newcal = Calendar()
newcal.add('prodid', '-//' + MY_NAME + '//' + MY_DOMAIN + '//')
newcal.add('version', '2.0')
newcal.add('x-wr-calname', CALENDARNAME)
DEBUGMSG = DEBUGMSG + 'new calendar started\n'

for s in glob.glob(CALDIR + '*.ics'):
   try:
      # open the file and read it
      calfile = open(s,'rb')
      cal = Calendar.from_ical(calfile.read())
      # every part of the file...
      for component in cal.subcomponents:
         if component.name == 'VEVENT':
	    # ...which name is VEVENT will be added to the new file
            newcal.add_component(component)
         if component.name == 'VTIMEZONE':
	    # ...which name is VEVENT will be added to the new file
            newcal.add_component(component)
      # close the existing file
      DEBUGMSG = DEBUGMSG + 'done with file ' + s +'\n'
      calfile.close()
   except:
      # if the file was not readable, we need a errormessage ;)
      print MY_SHORTNAME + ": Error: reading file:", sys.exc_info()[1]
      print s

# After the loop, we have all of our data and can write the file now
try:
   f = open(ICS_OUT, 'wb')
   f.write(newcal.to_ical())
   f.close()
   DEBUGMSG = DEBUGMSG + 'new calendar written\n'
except:
   print MY_SHORTNAME + ": Error: ", sys.exc_info()[1]

if DEBUG == True:
   try:
      l = open(DEBUGFILE, 'wb')
      l.write(DEBUGMSG)
      l.close()
   except:
      print MY_SHORTNAME + ": Error, unable to write log: ", sys.exc_info()[1]
# all done...
